stdlib/dict.spy

```diff
@@ -0,0 +1,202 @@
+from array import array1d
+# from hash import hash  # [Antonio]: I feel like this is supposed to work but doesn't
+from unsafe import gc_alloc, ptr
+
+
+@blue.generic
+def hash(T):
+    def hash_i32(v: i32) -> i32:
+        # this mimics the implementation of CPython's hash(int)
+        return v
+    if T is i32:
+        return hash_i32
+
+    raise TypeError("unsupported type for hash()")
+
+
+DKIX_EMPTY: i32 = 0
+# DKIX_DUMMY: i32 = 2 ** 32 - 1
+DKIX_DUMMY: i32 = 1 << 32 - 1
+# DKIX_ERROR: i32 = 2 ** 32 - 2
+DKIX_ERROR: i32 = 1 << 32 - 2
+# DKIX_KEY_CHANGED: i32 = 2 ** 32 - 3
+DKIX_KEY_CHANGED: i32 = 1 << 32 - 3
+
+MIN_LOG_SIZE: i32 = 6
+MAX_LOG_SIZE: i32 = 31
+MAX_FILL_RATIO: float = 2 / 3  # [Antonio]: should I use something other than float?
+
+
+@blue.generic
+def dict(Key, Value):
+    @struct
+    class Entry:
+        # empty: bool  # [Antonio]: a super-confusing error here:
+        #   ...
+        #   File "/Users/dp/repos/spy/spy/vm/modules/unsafe/struct.py", line 42, in calc_layout
+        #     field_size = sizeof(w_type)
+        #   File "/Users/dp/repos/spy/spy/vm/modules/unsafe/misc.py", line 22, in sizeof
+        #     assert False
+        #            ^^^^^
+        # AssertionError
+        empty: i32
+        key: Key
+        value: Value
+
+
+    @struct
+    class DictData:
+        index: ptr[array1d[i32]]
+        log_size: i32  # capacity
+        length: i32  # number of items stored
+        entries: ptr[array1d[Entry]]
+
+
+    def _new_data(log_size: i32) -> DictData:
+        # assert MIN_LOG_SIZE <= log_size <= MAX_LOG_SIZE
+        data = gc_alloc(DictData)(1)
+        index = array1d[i32](1 << log_size)
+        # index = array1d[i32](2 ** log_size)  # [Antonio]: missing ** operator
+        # also, the error was super confusing:
+        #   ...
+        #     w_OP = OP_from_token(binop.op) # e.g., w_ADD, w_MUL, etc.
+        #   File "/Users/dp/repos/spy/spy/vm/modules/operator/__init__.py", line 108, in OP_from_token
+        #     return _from_token[token]
+        #            ~~~~~~~~~~~^^^^^^^
+        # KeyError: '**'
+        data.index = index
```

Comment:
So I guess I need to change a pointer here, but how do I express that in spy?
Python certainly doesn't allow that, so I don't know the syntax off the top of my head